### PR TITLE
Verify POH in TVU

### LINKED ISSUE #1745: TVU does not verify PoH
#### Problem
validators do not verify that Entries are linked with PoH

#### Proposed Solution
add PoH verification to replicate_stage, initially with CPU, but later with GPU or enclave

### PR DESCRIPTION
**Pull request by sagar-solana**
#### Problem
Entries from blob fetch were not being verified

#### Summary of Changes
Propagated last_entry_id to tvu and replicate stage. Replicate will verify each entry recv'd against last_entry_id and keep last_entry_id updated as it goes.

Fixes #1745

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -16,3 +16,7 @@ Cargo.lock
 # log files
 *.log
 log-*.txt
+
```
Comment (rob-solana) [inline .gitignore]: good, but put this in a separate PR

File: src/entry.rs [hunk 2]
```diff
@@ -234,6 +235,25 @@ pub fn next_entry(start_hash: &Hash, num_hashes: u64, transactions: Vec<Transact
...
+    let mut num_ticks = 0;
+
+    for blob in blobs {
+        let entry: Entry = {
+            let msg = blob.read().unwrap();
+            let msg_size = msg.get_size()?;
+            deserialize(&msg.data()[..msg_size]).expect("Error reconstructing entry")
+        };
+
+        num_ticks += entry.is_tick() as u64;
```
Comment (rob-solana) [inline src/entry.rs]: uncast this, ok?

if entry.is_tick() { num_ticks += 1 }

File: src/fullnode.rs [hunk 3]
```diff
@@ -285,7 +286,7 @@ impl Fullnode {
                 ledger_path,
                 sigverify_disabled,
                 max_tick_height,
-                last_id,
+                last_entry_id,
```
Comment (rob-solana) [inline src/fullnode.rs]: really good naming

File: src/replicate_stage.rs [hunk 4]
```diff
@@ -136,7 +145,7 @@ impl ReplicateStage {
 
         *entry_height += entries_len;
         res?;
-        Ok(last_entry_id)
```
Comment (rob-solana) [inline src/replicate_stage.rs]: the return form is more canonical rust

mut is a four-letter word

File: src/replicate_stage.rs [hunk 5]
```diff
@@ -78,10 +80,17 @@ impl ReplicateStage {
         let mut res = Ok(());
         let last_entry_id = {
             let mut num_entries_to_write = entries.len();
+            let mut last_entry_id_ = *last_entry_id;
```
Comment (rob-solana) [inline src/replicate_stage.rs]: you can re-use the name if you like?

Comment (sagar-solana) [reply src/replicate_stage.rs]: I wanted some visible separation even if rust was ok with it.

**Comment by rob-solana** [comment]
would like to see before and after testnet-pref TPS report

File: .gitignore [hunk 6]
```diff
@@ -1,6 +1,5 @@
 Cargo.lock
 /target/
-/book/
```
Comment (mvines) [inline .gitignore]: why was this removed?

File: src/replicate_stage.rs (hunk 5 above)
Comment (rob-solana) [reply src/replicate_stage.rs]: ick ;)
